In [9]:
import numpy as np
import wget
import pandas as pd

# Question 1
**What's the version of NumPy that you installed?**

In [3]:
np.__version__

'1.20.3'

In [8]:
url = 'https://raw.githubusercontent.com/alexeygrigorev/mlbookcamp-code/master/chapter-02-car-price/data.csv'
wget.download(url)

100% [..........................................................................] 1475504 / 1475504

'data.csv'

In [10]:
data = pd.read_csv('data.csv')
data.head()

Make       Model  Year             Engine Fuel Type  Engine HP  \
0  BMW  1 Series M  2011  premium unleaded (required)      335.0   
1  BMW    1 Series  2011  premium unleaded (required)      300.0   
2  BMW    1 Series  2011  premium unleaded (required)      300.0   
3  BMW    1 Series  2011  premium unleaded (required)      230.0   
4  BMW    1 Series  2011  premium unleaded (required)      230.0   

   Engine Cylinders Transmission Type     Driven_Wheels  Number of Doors  \
0               6.0            MANUAL  rear wheel drive              2.0   
1               6.0            MANUAL  rear wheel drive              2.0   
2               6.0            MANUAL  rear wheel drive              2.0   
3               6.0            MANUAL  rear wheel drive              2.0   
4               6.0            MANUAL  rear wheel drive              2.0   

                         Market Category Vehicle Size Vehicle Style  \
0  Factory Tuner,Luxury,High-Performance      Compact         Coupe   
1                     Luxury,Performance      Compact   Convertible   
2                Luxury,High-Performance      Compact         Coupe   
3                     Luxury,Performance      Compact         Coupe   
4                                 Luxury      Compact   Convertible   

   highway MPG  city mpg  Popularity   MSRP  
0           26        19        3916  46135  
1           28        19        3916  40650  
2           28        20        3916  36350  
3           28        18        3916  29450  
4           28        18        3916  34500

# Question 2
**How many records are in the dataset?**

In [11]:
data.shape

(11914, 16)

11914 records

# Question 3
**Who are the most popular car manufacturers (top-3) according to the dataset?**

In [31]:
data.sort_values('Popularity').head()

Make Model  Year             Engine Fuel Type  Engine HP  \
2116   Spyker    C8  2009  premium unleaded (required)      400.0   
2117   Spyker    C8  2009  premium unleaded (required)      400.0   
2118   Spyker    C8  2009  premium unleaded (required)      400.0   
5202  Genesis   G80  2017             regular unleaded      311.0   
5204  Genesis   G80  2017             regular unleaded      311.0   

      Engine Cylinders Transmission Type     Driven_Wheels  Number of Doors  \
2116               8.0            MANUAL  rear wheel drive              2.0   
2117               8.0            MANUAL  rear wheel drive              2.0   
2118               8.0            MANUAL  rear wheel drive              2.0   
5202               6.0         AUTOMATIC  rear wheel drive              4.0   
5204               6.0         AUTOMATIC   all wheel drive              4.0   

              Market Category Vehicle Size Vehicle Style  highway MPG  \
2116  Exotic,High-Performance      Compact         Coupe           18   
2117  Exotic,High-Performance      Compact   Convertible           18   
2118  Exotic,High-Performance      Compact         Coupe           18   
5202       Luxury,Performance        Large         Sedan           28   
5204       Luxury,Performance        Large         Sedan           25   

      city mpg  Popularity    MSRP  
2116        13           2  209990  
2117        13           2  219990  
2118        13           2  209990  
5202        18          21   41400  
5204        16          21   43900

Top 3 most popular car manufacturers is Chevrolet, Ford, Toyota

# Question 4
**What's the number of unique Audi car models in the dataset?**

In [34]:
data[data.Make == 'Audi'].Model.nunique()

34

34 unique Audi car models

# Question 5
**How many columns in the dataset have missing values?**

In [44]:
data.isnull().sum()

Make                    0
Model                   0
Year                    0
Engine Fuel Type        3
Engine HP              69
Engine Cylinders       30
Transmission Type       0
Driven_Wheels           0
Number of Doors         6
Market Category      3742
Vehicle Size            0
Vehicle Style           0
highway MPG             0
city mpg                0
Popularity              0
MSRP                    0
dtype: int64

5 Columns

# Question 6
* Find the median value of "Engine Cylinders" column in the dataset.
* Next, calculate the most frequent value of the same "Engine Cylinders".
* Use the fillna method to fill the missing values in "Engine Cylinders" with the most frequent value from the previous step.
* Now, calculate the median value of "Engine Cylinders" once again.

In [45]:
data['Engine Cylinders'].median()

6.0

In [51]:
data['Engine Cylinders'].mode()

0    4.0
dtype: float64

In [56]:
data['Engine Cylinders'] = data['Engine Cylinders'].fillna(4)

In [57]:
data['Engine Cylinders'].median()

6.0

No, Median has not changed

# Question 7
* Select all the "Lotus" cars from the dataset.
* Select only columns "Engine HP", "Engine Cylinders".
* Now drop all duplicated rows using drop_duplicates method (you should get a dataframe with 9 rows).
* Get the underlying NumPy array. Let's call it X.
* Compute matrix-matrix multiplication between the transpose of X and X. To get the transpose, use X.T. Let's call the result XTX.
* Invert XTX.
* Create an array y with values [1100, 800, 750, 850, 1300, 1000, 1000, 1300, 800].
* Multiply the inverse of XTX with the transpose of X, and then multiply the result by y. Call the result w.
* What's the value of the first element of w?

In [65]:
data[data.Make == 'Lotus'].head()

Make  Model  Year             Engine Fuel Type  Engine HP  \
3912  Lotus  Elise  2009  premium unleaded (required)      189.0   
3913  Lotus  Elise  2009  premium unleaded (required)      218.0   
3914  Lotus  Elise  2009  premium unleaded (required)      189.0   
3915  Lotus  Elise  2010  premium unleaded (required)      189.0   
3916  Lotus  Elise  2010  premium unleaded (required)      218.0   

      Engine Cylinders Transmission Type     Driven_Wheels  Number of Doors  \
3912               4.0            MANUAL  rear wheel drive              2.0   
3913               4.0            MANUAL  rear wheel drive              2.0   
3914               4.0            MANUAL  rear wheel drive              2.0   
3915               4.0            MANUAL  rear wheel drive              2.0   
3916               4.0            MANUAL  rear wheel drive              2.0   

              Market Category Vehicle Size Vehicle Style  highway MPG  \
3912  Exotic,High-Performance      Compact   Convertible           27   
3913  Exotic,High-Performance      Compact   Convertible           26   
3914  Exotic,High-Performance      Compact   Convertible           27   
3915  Exotic,High-Performance      Compact   Convertible           27   
3916  Exotic,High-Performance      Compact   Convertible           26   

      city mpg  Popularity   MSRP  
3912        21         613  43995  
3913        20         613  54990  
3914        21         613  47250  
3915        21         613  47250  
3916        20         613  54990

In [66]:
data[data.Make == 'Lotus'][[ "Engine HP", "Engine Cylinders"]].head()

Engine HP  Engine Cylinders
3912      189.0               4.0
3913      218.0               4.0
3914      189.0               4.0
3915      189.0               4.0
3916      218.0               4.0

In [67]:
data[data.Make == 'Lotus'][[ "Engine HP", "Engine Cylinders"]].drop_duplicates()

Engine HP  Engine Cylinders
3912      189.0               4.0
3913      218.0               4.0
3918      217.0               4.0
4216      350.0               8.0
4257      400.0               6.0
4259      276.0               6.0
4262      345.0               6.0
4292      257.0               4.0
4293      240.0               4.0

In [70]:
X = data[data.Make == 'Lotus'][[ "Engine HP", "Engine Cylinders"]].drop_duplicates().values
X

array([[189.,   4.],
       [218.,   4.],
       [217.,   4.],
       [350.,   8.],
       [400.,   6.],
       [276.,   6.],
       [345.,   6.],
       [257.,   4.],
       [240.,   4.]])

In [75]:
XTX = X.T.dot(X)
XTX

array([[7.31684e+05, 1.34100e+04],
       [1.34100e+04, 2.52000e+02]])

In [78]:
inv_XTX = np.linalg.inv(XTX)
inv_XTX

array([[ 5.53084235e-05, -2.94319825e-03],
       [-2.94319825e-03,  1.60588447e-01]])

In [77]:
y = np.array([1100, 800, 750, 850, 1300, 1000, 1000, 1300, 800])
y

array([1100,  800,  750,  850, 1300, 1000, 1000, 1300,  800])

In [79]:
w = inv_XTX.dot(X.T).dot(y)
w

array([  4.59494481, -63.56432501])

the first element of w is 4.5949